### PR TITLE
Prepare UI Context and Registration

### PR DESCRIPTION
**Pull request by VictorS67**

### CODE REVIEW

File: packages/app/src/types/studio.type.ts
Comment (AlveinTeng) [inline packages/app/src/types/studio.type.ts:1]: Can you add some explanatory comment about each BuiltInNodeType to make it easier for the developers from community

Comment (VictorS67) [reply packages/app/src/types/studio.type.ts:1]: Please ignore the code update in `encre-app` for now. The full update should be in other pull requests.

**Comment by VictorS67** [comment]
# Node Registration System

## Overview
The Node Registration System is designed to facilitate the dynamic creation and management of a variety of nodes in the Encre Studio. It allows for the registration, creation, and retrieval of node instances based on type and subtype, supporting a flexible architecture that can accommodate various functionalities. This system is crucial for developers looking to extend the project with new node types or leverage existing nodes for building complex features.

## Key Concepts
- **Node**: A fundamental unit of functionality, represented by a class that implements the `SerializableNode` interface.
- **Node Types and SubTypes**: Each node has a `type` and an optional `subType` to uniquely identify its purpose and capabilities.
- **Node Registration**: The process of adding a node class to the system, enabling it to be instantiated and used within the application.
- **Node Creation**: The process of dynamically instantiating nodes based on their type and subtype.

## Registering Nodes
Nodes are registered through the `NodeRegistration` class, which maintains a registry of node types, their implementations, and associated metadata. The system supports the registration of built-in nodes as well as custom nodes developed by contributors.

### Built-in Nodes
The system comes pre-loaded with several built-in nodes, including:

- Loader nodes (e.g., `PDFLoaderNodeImpl`) for handling various input formats.
- Message nodes (e.g., `HumanMessageNodeImpl`, `BotMessageNodeImpl`) for representing different types of messages within the system.
- Prompt nodes (e.g., `StringPromptNodeImpl`, `ChatPromptNodeImpl`) for managing user input prompts.
- Splitter nodes (e.g., `TextSplitterNodeImpl`, `LanguageTextSplitterNodeImpl`) for text segmentation and processing.
- Inference nodes (e.g., `OpenAINodeImpl`, `GeminiNodeImpl`) for integrating machine learning models and algorithms.

### Registering Custom Nodes
To register a custom node, developers must extend the `SerializableNode` interface and implement the required functionality. The custom node is then registered using the `register` method of the `NodeRegistration` class, specifying the node implementation and any additional arguments needed for instantiation.

## Creating Nodes
Nodes can be created using the `create` method of the `NodeRegistration` instance, providing the type and subtype of the desired node along with any necessary arguments. This method dynamically instantiates the node, allowing for flexible use within the Encre Studio.

## Example Usage
```typescript
// Define a custom node
class MyCustomNodeImpl extends SerializableNode {
  // Implementation details
}

// Register the custom node
globalNodeRegistry.register(MyCustomNodeImpl);

// Create an instance of the custom node
const myCustomNode = globalNodeRegistry.create('MyCustomNodeType', 'SubType', { /* arguments */ });
```

## Resetting the Node Registry
The `resetNodeRegistry` function can be used to reset the global node registry, clearing all registered nodes and allowing for a fresh start.

**Comment by VictorS67** [comment]
# UI Context Handling in Node Body

## Overview
This document introduces the UI Context Handling in Node Body, a feature designed to map various data types to specific UI contexts for rendering the node body in the Encre Studio. This enables a dynamic and flexible way to display data based on its type, providing user interaction and the visual presentation of information.

## UI Context Types
There are several UI context types, each specific to present data in a way that best suits its nature. These contexts include plain text, markdown, code blocks, binary blobs, contextual information, and messages. Additionally, a base UI context provides common settings like font size and family, which can be extended by other contexts.

### BaseUIContext
- **Attributes:** `fontSize`, `fontFamily`
- **Purpose:** Sets basic styling options for UI elements. This is shared to all UI Contexts.

### PlainUIContext
- **Attributes:** `type`, `text`
- **Purpose:** Displays simple text without formatting.

### MarkdownUIContext
- **Attributes:** `type`, `text`
- **Purpose:** Renders text with Markdown formatting for rich text presentation.

### CodeUIContext
- **Attributes:** `type`, `text`, `language`, `keywords`
- **Purpose:** Shows code snippets with optional syntax highlighting based on the specified language and keywords. The [Monaco Editor](https://microsoft.github.io/monaco-editor/) is used in visualizing the code block.

### BlobUIContext
- **Attributes:** `type`, `blob`, `size`, `blobType`
- **Purpose:** Handles binary data, such as files, showing metadata like size and type.

### ContextUIContext
- **Attributes:** `type`, `text`, `metadata`
- **Purpose:** Displays contextual information with optional metadata for detailed insights.

### MessageUIContext
- **Attributes:** `type`, `text`, `role`, `name`, `additionalKwargs`
- **Purpose:** Represents messages, supporting roles, sender names, and additional keyword arguments about other information of the messages.

## UI Context Mapping
The system employs a mapping strategy (`UIDataTypesMap`) to associate data types with their corresponding UI contexts. This mapping ensures that data is presented in the most appropriate format, improving readability and user experience.

### UIDataTypesMap
This record maps `DataType` (derived from various data types like string, number, boolean, etc.) to `UIContext['type']`, determining how data should be rendered in the UI. For example, code-related data types are mapped to the `code` UI context, while binary blobs are linked to the `blob` context.

- Note: `plain`, `markdown`, and `code` contexts are used for all primitive data types (e.g. `string`, `number`, `boolean`, `object`, `unknown`, and their array types). Usually, a `markdown` context is not directly used for rendering the data, but we allow a prefix of `::markdown` in the string to render the current string data into a markdown context.

## Usage
To utilize the UI Context Handling, developers should:

1. **Identify the Data Type:** Determine the type of data to be displayed in the UI.
2. **Map to UI Context:** Use `UIDataTypesMap` to find the corresponding UI context for the data type.
3. **Render UI Context:** Implement the UI rendering logic based on the identified UI context, customizing as needed for the application's design and user experience requirements.

**Comment by VictorS67** [comment]
# Dynamic UI Displaying Utils

## Overview
This documentation describes the Dynamic UI Display Utils to dynamically generate user interface elements of the node body based on the structured data provided.

## Components

### Data Fields (`DataFields`)
`DataFields` is a type that represents a collection of data entries, where each entry is associated with a unique key. This structure is used for storing the keyword arguments of a node, which should be displayed in the UI.

### Secret Fields (`SecretFields`)
`DataFields` is a type that represents a collection of secret keys stored in a configuration file and the corresponding name in a node, which should be displayed in the UI.

### UI Contexts (`UIContext`)
`UIContext` types define various ways data can be represented in the UI. These contexts include, but are not limited to, plain text, code blocks, markdown content, binary blobs, contextual information, and messages.

### Display Functions
There are two key functions to process data fields and generate corresponding UI contexts:

- `displayUIFromDataFields(dataFields: DataFields): UIContext[]`: Generating UI Contexts from the given Data Fields.
- `displayUIFromSecretFields(secrets: SecretFields): UIContext[]`: Generating UI Contexts from the given Secret Fields.

## Workflow

1. **Data Partitioning:** Data fields are first partitioned based on their UI context type, ensuring that data types are grouped appropriately for display.

2. **UI Context Generation:** For each partitioned group, the system generates a UI context (or multiple contexts) that best represents the data. This process involves matching data types to their respective UI representations and handling special cases like concatenation of similar context types for a cohesive display.

3. **Dynamic UI Rendering:** The generated UI contexts are then ready to be rendered in the application's user interface, allowing for a dynamic and adaptive presentation of the underlying data.

## Key Functions

### `displayUIFromDataFields`
This function takes `DataFields` as input and generates an array of `UIContext` elements. It uses pattern matching to determine the appropriate UI representation for each data type, supporting a rich and flexible UI composition.

### `displayUIFromSecretFields`
Specifically designed to handle secret fields, this function generates code UI contexts that obscure sensitive information, suitable for displaying secrets or confidential data securely.

## Usage

To utilize or extend the Dynamic UI Displaying, developers should:

- **Data Structure:** Ensure that data provided to the display functions is structured according to the `DataFields` type specification.
- **Custom UI Contexts:** Developers may extend or customize UI contexts to better suit specific data presentation needs, adhering to the established patterns for consistency and compatibility.
- **Performance Considerations:** Be mindful of the performance implications when processing large datasets or complex data structures, optimizing where necessary.

File: packages/app/src/types/studio.type.ts
Comment (VictorS67) [reply packages/app/src/types/studio.type.ts:1]: I've added rendering UI for plain text, markdown, and code UI context type.

**Comment by VictorS67** [comment]
I've added rendered UI for plain text, markdown, and code UI context type:

Plain text UI context:
firefox_jj5gmzRQBc

markdown UI context:
firefox_wRsm88byQr

code UI context:
firefox_9tqp9p05hy

**Review by AlveinTeng — APPROVED** [review]

**Review by AlveinTeng — APPROVED** [review]

**Review by AlveinTeng — APPROVED** [review]

**Review by BenWXY — APPROVED** [review]

**Pull request merged**